src/Microsoft.AspNetCore.OData/Batch/ODataBatchReaderExtensions.cs

```diff
@@ -236,5 +268,103 @@ private static HttpContext CreateHttpContext(HttpContext originalContext)
+        /// If a given preference is defined in both the batch and individual request, the one from the individual
+        /// request is retained and the one from the batch is discarded.
+        /// </summary>
+        /// <param name="individualPreferences">The value of the Prefer header from the individual request inside the batch</param>
+        /// <param name="batchPreferences">The value of the Prefer header from the overall batch request</param>
+        /// <returns>Value containing the combined preferences</returns>
+        private static string MergeIndividualAndBatchPreferences(string individualPreferences, string batchPreferences)
+        {
+            if (string.IsNullOrEmpty(individualPreferences))
+            {
+                return batchPreferences;
+            }
+            if (string.IsNullOrEmpty(batchPreferences))
+            {
+                return individualPreferences;
+            }
+            // get the name of each preference to avoid adding duplicates from batch
+            var individualList = individualPreferences.SplitPreferences().Select(pref => pref.Trim());
+            var individualPreferenceNames = new HashSet<string>(individualList.Select(pref => pref.Split('=').FirstOrDefault()));
+            
+            
+            var filteredBatchList = batchPreferences.SplitPreferences().Select(pref => pref.Trim())
+                // do not add duplicate preferences from batch
+                .Where(pref => !individualPreferenceNames.Contains(pref.Split('=').FirstOrDefault()));
+            var filteredBatchPreferences = string.Join(", ", filteredBatchList);
+
+            return string.Join(", ", individualPreferences, filteredBatchPreferences);
+        }
+
+        /// <summary>
+        /// Splits the value of a Prefer header into separate preferences
+        /// e.g. a value like 'a, b=c, foo="bar,baz"' will return an IEnumerable with
+        /// - a
+        /// - b=c
+        /// - foo="bar,baz"
+        /// </summary>
+        /// <param name="preferences"></param>
+        /// <returns></returns>
+        private static IEnumerable<string> SplitPreferences(this string preferences)
+        {
+            StringBuilder currentPreference = new StringBuilder();
+            var insideQuotedValue = false;
+            char prevChar = '\0';
+            foreach (var c in preferences)
+            {
+                if (c == '"')
+                {
+                    if (!insideQuotedValue)
+                    {
+                        // we are starting a double-quoted value
+                        insideQuotedValue = true;
+                    }
+                    else
+                    {
+                        // this could be the end of a quoted value, or it could be an escaped quote
+                        insideQuotedValue = prevChar == '\\';
+                    }
+                }
+
+                if ((c == ',' || c == ';') && !insideQuotedValue)
```

Comment:
>(c == ',' || c == ';' [](start = 20, length = 21)

should just be ;?

Reply:
You're right. I misinterpreted the grammar, the semicolon is just used to separate parameters of a given preference, not to separate preferences. I will correct that.

Reply:
I also had to re-read; I initially thought that semi-colon was the separator and was wondering at the use of comma, until I looked more closely and realized it was the other way around.

---
In reply to: [384954431](https://github.com/OData/WebApi/pull/2064#discussion_r384954431) [](ancestors = 384954431)